Cup Products, stage 2

timjb opened the pull request.

timjb committed:
Define cup product (in degrees m=n=1) on smash product

timjb committed:
Refactor smash product, define ∧-fmap

timjb committed:
Working towards proposition about connectivity of ∧-fmap

This is based on subsection 4.3 from Guillaume Brunerie's thesis

favonia commented on core/lib/types/Smash.agda:
```diff
@@ -6,13 +6,11 @@ open import lib.types.Coproduct
-open import lib.types.Wedge
 
-module lib.types.Smash {i j} where
+module lib.types.Smash where
```
@timjb Is this due to some type-checking issue or just coding style (which is fine)? It seems all the following still has "{i j}" as the prefix.

timjb replied:
That's a leftover of some development that I did in this file (which required generalizing the universe levels) which I later moved to a separate (still uncommitted) file. You're right that this change is not needed (anymore). Still, I like the way it is now better: Introducing the universe levels together with the types they're for seems cleaner to me.

favonia replied:
I actually agree! I was just curious if there's any other reason. :smile:

timjb committed:
Filled a hole in wip proof of proposition about connectivity of ∧-fmap

timjb committed:
Finish first proposition about ∧-fmap connectivity

timjb committed:
Prove symmetric variant of ∧-fmap connectivity proposition

timjb committed:
Combined proposition about connectivity of ∧-fmap

timjb committed:
Σ^n (EM₁ G) ∧ Σ^m (EM₁ H) → EM G (S n) ∧ EM H (S m) is (Sn+Sm)-connected

where both components in the map are induced by truncation

timjb committed:
Define cup product map EM G (S n) ∧ EM H (S n) → EM (G ⊗ H) (S n + S m)

timjb committed:
Towards commutativity of cup product in higher degrees

timjb committed:
Relation between ∧Σ-out, Σ∧-out and ∧-swap

timjb committed:
Commutativity of ∧Σ^-out and Σ^∧-out

timjb committed:
Relation between Σ^∧Σ^-out and ∧-swap

favonia commented on core/lib/types/Nat.agda:
```diff
@@ -77,6 +78,13 @@ instance
+
+odd : ℕ → Bool
+odd = negate ∘ even
+
```
@timjb I wonder `even (S n) = odd n` and `odd (S n) = even n` would be easier because of the double negation in the following code. I could be wrong.

timjb replied:
@favonia One advantage of the definition `odd = negate ∘ even` is that `odd n`, `even n`, `odd (S n)`, `even (S n)`, ... all reduce to `even n`. With `even` and `odd` defined by mutual recursion, there are two propositions `even n` and `odd n` for every `n` that are not judgmentally related. So you would have to prove (by induction) that e.g. `odd n == negate (even n)` and `even n == negate (odd n)`. With the current definitions of `even` and `odd`, the first path is reflexivity and the second path is an instance of `!-!`.

timjb committed:
Prove commutativity of cup product in degrees n, m >= 1

timjb committed:
Define cup product on EM in degrees m, n >= 0

timjb committed:
Prove commutativity of cup product in all degrees

timjb committed:
Define cup product on cohomology

timjb committed:
Cosmetic changes to IterSuspSmash

timjb committed:
Prove commutativity of the cup product on cohomology groups

This required extending the commutativity proof of the cup product on
Eilenberg--MacLane spaces from a path between functions to a path
between pointed functions.

timjb committed:
Split (iterated) suspension code into more modules

timjb commented:
@favonia I've finally finished proving commutativity of the CP defined on cohomology groups! This has taken a lot more time that I expected, since it required going over the whole proof of commutativity of the CP on Eilenberg--MacLane spaces and generalizing everything from normal maps to pointed maps.

Beginning tomorrow, I will stop formalizing new stuff and resume writing everything up. I would appreciate it if you could review this PR sometime this month or next month before merging it.

timjb committed:
Replace Σ∧-out-smin and ∧Σ-out-smin by Susp-fmap

favonia commented on core/lib/PathFunctor.agda:
```diff
@@ -302,46 +311,51 @@ module _ {i} {A : Type i} where
     (h : (x : A) → f x == b) {x y : A} (p : x == y)
     → ap f p == h x ∙ ! (h y)
-  homotopy-naturality-to-cst f b h {x} p@idp = ! (!-inv-r (h x))
+  homotopy-to-cst-ap f b h {x} p@idp = ! (!-inv-r (h x))
```
What does the ending `ap` mean? It seems `naturality` also involves `ap`. I am not against this, but I would like to learn the rationale.

timjb replied:
I chose the suffix `-ap` because the statement involves `ap`. I'm not happy with this choice either. Two suggestions:

* `homotopy-naturality-to-cst'` would make sense since the statement is similar to ``homotopy-naturality-to-cst` (with the paths "rotated").
* `ap-homotopic-to-cst` would make sense, since the definition can be seen as the generalization of `ap-cst` where the function is not judgmentally the constant function, but only homotopic to it. (This is my favorite.)

favonia replied:
Sorry for the late reply. I am more worried about consistency, and I think both are compatible with the documented convention. (In fact, I think the intention was to adopt cubical methods instead of naturality lemmas.) I wonder how one can generalize the `homotopic` so that it can be registered as a general principle.

timjb replied:
Another proposal: `ap-null-homotopic`

Citing [Wikipedia](https://en.wikipedia.org/wiki/Homotopy#Null-homotopy):

> A function f is said to be null-homotopic if it is homotopic to a constant function.

I am not sure what kind of generalization you are seeking. In our naming convention, `null-homotopic` (or `homotopic-to-cst`) could be used everywhere where `cst` or `idf` is used.

favonia replied:
I was worried about the parsing of a name involving `to`. In general, `A-to-B` is the (canonical) function from `A` to `B`. In this case, it is fine because `cst` is not a type, but I do not know what to say if we are going to consider `homotopic-to-X` for generic `X`. `null-homotopic` also works because there is no confusion in special cases.

timjb replied:
I've renamed the function to `ap-null-homotopic`.

favonia commented on core/lib/types/Group.agda:
```diff
@@ -289,6 +302,12 @@ contr-is-trivialᴳ : ∀ {i} (G : Group i)
+  → AbGroup.grp G == AbGroup.grp H
+  → G == H
+abgroup= G H p =
+  pair= p (contr-center (↓-level (is-abelian-is-prop (AbGroup.grp H))))
```
You can probably use `Subtype=` here.

timjb replied:
Yes, that works!

favonia commented on core/lib/groups/TensorProduct.agda:
```diff
@@ -43,11 +43,37 @@ module TensorProduct₀ {i} {j} (G : AbGroup i) (H : AbGroup j) where
+    ⊗-lin-r g h₁ h₂ = rel-holds (linear-r g h₁ h₂)
+
+    ⊗-ident-l : ∀ h → G.ident ⊗ h == ident
+    ⊗-ident-l h = cancel-l (G.ident ⊗ h) $
```
I believe we name this lemma as `unit-l` in other places.

timjb replied:
That's right. I will change that.

favonia commented on core/lib/groups/TensorProduct.agda:
```diff
@@ -43,11 +43,37 @@ module TensorProduct₀ {i} {j} (G : AbGroup i) (H : AbGroup j) where
+      comp (G.ident ⊗ h) ident =∎
+
+    ⊗-ident-r : ∀ g → g ⊗ H.ident == ident
+    ⊗-ident-r g = cancel-r (g ⊗ H.ident) $
```
And `unit-r` for this one.

favonia commented on theorems/cohomology/CupProduct/Definition.agda:
```diff
@@ -0,0 +1,304 @@
+  _∪_ {m} {n} s t = Trunc-rec (λ s' → Trunc-rec (λ t' → [ ⊙Ω∧-cp m n ⊙∘ smin-map s' t' ]) t) s
+
+  -- ∪-hom : ∀ {m n : ℕ} → TensorProduct.grp (C-abgroup G (pos m) X) (C-abgroup H (pos n) X) →ᴳ C G⊗H.abgroup (pos (m + n)) X
+  -- ∪-hom = {!!}
```
Is this something you are planning to do in the future?

timjb replied:
Currently I'm not planning to do this. I have already deleted this comment.

favonia reviewed (approved):
This is such an amazing work! Thanks for the contribution.

timjb committed:
Simplify definition of cup product by avoiding unnecessary dualization

timjb committed:
Remove dead dual 2-semi-categories code

timjb committed:
small cleanups

timjb committed:
small cleanups

timjb committed:
clean up: Π-transp is only a special case of transp-naturality

timjb committed:
small simplifications

timjb committed:
Refactor: define Eilenberg--MacLane cup product on ⊙× instead of ⊙∧

This suffices for defining the cup product on reduced cohomology and
makes the proof of commutativity easier. This commit also makes the
definitions of the cup product in degrees (0, 1+k) and (1+k, 0) more
symmetric.

timjb committed:
Cohomology H^n(X; G) is functorial in the abelian group G

timjb committed:
Refactor: extract proof that inv-hom induces group inverse map in cohom.

timjb committed:
simplify ⊙Ω-!-Susp-flip'

timjb committed:
small changes

timjb committed:
Refactor: small details suggested by @favonia

timjb committed:
Rename 'homotopy-to-cst-ap' to 'ap-null-homotopic'

timjb commented:
@favonia I think this is ready to merge!

favonia commented:
@timjb Do you want to pull the trigger?

Alizter commented:
This is a bit daunting to go through. Could you write a small summary of what has been added?

favonia commented:
@timjb Maybe it would be a good idea to update `README.md` with short descriptions of the new files and pointers to your thesis. :thinking:

timjb commented:
@favonia Good idea! I will do this.
